Core.ApplicationServices/References/ReferenceService.cs

```diff
@@ -200,22 +200,43 @@ private void RaiseRootUpdated(IEntityWithExternalReferences owner)
-
-                            if (result.Failed)
-                                return new OperationError($"Failed to add reference with data:{JsonConvert.SerializeObject(referenceProperties)}. Error:{result.Error.Message.GetValueOrEmptyString()}", result.Error.FailureType);
+                            ExternalReference externalReference;
+                            if (externalReferenceProperties.Uuid.HasValue)
+                            {
+                                var uuid = externalReferenceProperties.Uuid.Value;
+                                var existingReferenceResult = _referenceRepository.GetByUuid(uuid);
+                                if(existingReferenceResult.IsNone)
+                                    return new OperationError($"External reference with uuid: {uuid} was not found", OperationFailure.NotFound);
 
-                            if (referenceProperties.MasterReference)
+                                externalReference = existingReferenceResult.Value;
+                                MapExternalReference(externalReferenceProperties, externalReference);
+                                referencesToUpdate.Add(externalReference);
+                            }
+                            else
                             {
-                                var masterReferenceResult = root.SetMasterReference(result.Value);
-                                if (masterReferenceResult.Failed)
-                                    return new OperationError($"Failed while setting the master reference:{masterReferenceResult.Error.Message.GetValueOrEmptyString()}", masterReferenceResult.Error.FailureType);
+                                var addReferenceResult = AddReference(rootId, rootType, externalReferenceProperties.Title, externalReferenceProperties.DocumentId, externalReferenceProperties.Url);
+
+                                if (addReferenceResult.Failed)
+                                    return new OperationError($"Failed to add reference with data:{JsonConvert.SerializeObject(externalReferenceProperties)}. Error:{addReferenceResult.Error.Message.GetValueOrEmptyString()}", addReferenceResult.Error.FailureType);
+
+                                externalReference = addReferenceResult.Value;
                             }
+
+                            if (!externalReferenceProperties.MasterReference) 
+                                continue;
+
+                            var masterReferenceResult = root.SetMasterReference(externalReference);
+                            if (masterReferenceResult.Failed)
+                                return new OperationError($"Failed while setting the master reference:{masterReferenceResult.Error.Message.GetValueOrEmptyString()}", masterReferenceResult.Error.FailureType);
                         }
+
+                        _referenceRepository.UpdateRange(referencesToUpdate);
```

Comment:
no need to maintain that list and call that method.. the save should go to the root which owns the references right?